---
title: "NPS -- Adopting tipping-point theory to transcriptome profiles to unravel disease regulatory trajectory"
author: "Zhezhen Wang, Biniam Feleke and Xinan H.Yang"
date: "`r format(Sys.Date(), '%m/%d/%Y')`"
abstract: >
  Network perturbation signature (NPS). The purpose of this R package NPS, is to
  analyze clinical transcriptome data in order to determine critical transition
  from phenotype-defined sample states. Fundamental of the tipping-point is
  applied to transcriptomic profiles where highly connected transcriptional
  configurations serve as attractors [1]. NPS improves on a method called
  Dynamic Network Biomarker (DNB) that was calculated within each state [2].
  When studying NB, expressional patterns are often the aggregates of multiple
  dynamic subsystems and thus a new method is needed to allow for cross-state
  comparisons. This makes it vague to use DNB for tipping-point detection
  because the autocorrelation of non-DNB-modules at other states could attribute
  to an even higher DNB-score at a predicted ‘tipping point’ (Figure 1). By
  incorporating an index of critical state transition (IC-score) [3] into the
  DNB model, NPS overcomes this pitfall by checking for  drastic deviations
  across samples around the tipping point. Thus, the NPS scheme acts as a hybrid
  model to join the advantages of both DNB and IC-scoring.
output:  
  word_document: default
  pdf_document: default
  html_document: 
    fig_caption: yes
    keep_md: yes
    toc: yes
vignette: >
  %\VignetteIndexEntry{title}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
---

#### ####

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, echo=FALSE, fig.cap="Fig 1. NPS workflow with five key analytic steps", fig.align='center', out.width = '60%'}
knitr::include_graphics("Fig1.jpg")
```

#### [Standard workflow](#Standard workflow)
   * ##### [Transcript Annotation and Biotype](#Transcript Annotation and Biotype)
     *  ##### [Quick Start](#Quick Start)
     *  ##### [Genomic Data Source](#Genomic Data Source)
     *  ##### [Extracting Summary Data](#Extracting Summary Data)
     *  ##### [Loading Data](#Loading Data)
     *  ##### [Filtering chr21](#Filtering chr21)
   * ##### [An Identification of Critical Tipping Point](#An Identification of Critical Tipping Point)
     *  ##### [Data Preprocessing](#Data Preprocessing)
     *  ##### [Pre-selection Transcript](#Pre-selection Transcript)
     *  ##### [Network Partition](#Network Partition)
     *  ##### [Identifying Dynamic Network Biomodule](#Identifying Dynamic Network Biomodule)
     *  ##### [Finding Tipping Point](#Finding Tipping Point)
   * ##### [Acknowledgements](#Acknowledgements)
   * ##### [SessionInfo](#SessionInfo)
   * ##### [References](#References)
<a name="Transcript Annotation and Biotype"></a>

#### __Transcript Annotation and Biotype__
<a name="Quick Start"></a>

#### __Quick Start__

  The R function getBiotype is used to group transcripts of interest into 11
  biotypes based on GENCODE annotation (Fig 2a). When a query transcript
  overlaps at least half with a GENCODE transcript on the same strand, this
  query transcript will inherit the biotype of the GENCODE transcript.

  In the previous study conducted, five out of the 11 biotypes showed high
  protein-coding potential while the others did not (Fig 2b) [4]. We thus
  concluded these seven other biotypes, including protein-coding antisense RNAs,
  to be lncRNAs. The remaining coding biotypes in this study included canonic
  protein coding (CPC), ‘PC_mixed’, and ‘PC_intron’.

  First start by loading the required libraries: “GenomeInfoDb,” “NPS,”
  “GenomicRanges,” “IRanges” and “NPS”. Next load the datasets: “gencode”,
  “ILEF”, “intron” and “cod”. Using these datasets, excute NPS functions
  getBiotypes and getReadthrough as follows. These steps assume you installed
  the “NPS” package. If you did not install the package, use the
  `install.packages("NPS")` to install in R.

```{r, echo=FALSE, fig.align='center', out.width = '65%'}
knitr::include_graphics("Fig2.jpg")
```
Fig 2. A getBiotypes workflow and protein-coding potential in real data analysis
[4]. (a) Workflow of an in-house R function (getBiotypes) to query transcripts
of interests and classify into biotypes. (b) Pie-chart of eleven types of
transcripts assembled from polyadenylated RNA(TARGET). (c) Empirical cumulative
distribution plot comparing the transcripts across all 11 biotypes. The
protein-coding potential was estimated with the Coding Potential Assessment Tool
(CPAT). Line color codes biotypes. The more a line towards the right-bottom
corner, the lower protein-coding potential it has.

```{r}
#library("GenomeInfoDb")
#library("GenomicRanges")
#library("IRanges")
#library("NPS")
#data("gencode")
#head(gencode)
```

  These illustrations above assumes you have installed "NPS" package. If you did
  not install the pacakge, use the `install.packages("NPS")` to install in R.

<a name="Genomic Data Source"></a>\

#### __Genomic Data Source__ 

  High quality human genome sequence data can be obtained from various sources.
  We obtained a comprehensive gene annotation of human GRCh37 from
  [GENCODE](https://www.gencodegenes.org/human). For our illustrations, human
  GRCh37 data will be used. A standard file structure, similar to gtf format, is
  required for this package. The general transfer format (gtf) organizes genomic
  data in rows and columns (fields). Each row contains information about
  specific samples. The columns are tab separated headers of the data
  frame.There are eight fixed columns with specific headers. An example of gtf
  format is shown below. For details of the gtf file format visit this
  [link](https://useast.ensembl.org/info/website/upload/gff.html#tracklines
  target="_blank").


   ![Chromosome 21 of human GRCh37 gtf](chr21.jpg)

  The table above shows head of chr21 dataset which was extracted from a full
  genome dataset GRCh37. The extraction of subset Chr21 is described below.

<a name="">[GoTop](#)</a> 
<a name="Extracting Summary Data"></a>\

##### __Extracting Summary Data__

  Before any further analysis, we need to summarize the content of the raw gtf
  data. There are two ways to get genome biotypes: a) "transctipt_type" b)
  "gene_type". Due to our interst in coding and noncoding regions, the
  `transcript_type` method was used to extract the regions of interest using
  python script shown below. __Note__ that the `"PATH_FILE"` refers to the path
  where the downloded gtf file is located. For instance, if the gtf file is
  located on your `desktop`, replace the `"PATH_FILE"` with
  `"Users/user/Desktop/gtf"`.

**Python codes:**

```{r "python code", eval = FALSE}
gtf = ("PATH_FILE")
outF = open("gtf_summary_transbiotype.txt","w")

def getquote(str,f,target):
    targetLen = len(target)+2
    strInd = str.find(target)
    st = strInd + len(target)+2
    ed = st + str[st:].find("";")
    #print(st,ed)
    f.write("\t"+str[st:ed]) if strInd!= -1 else f.write("\t"+"NA.")

with open(gtf, "r") as f:
     for line in f:
        if line[0] != "#":
            chromosome = line.split("\t")[0]
            st = line.split("\t")[3]
            ed = line.split("\t")[4]
            strand = line.split("\t")[6]
            type = line.split("\t")[2]
            outF.write(chromosome+"\t"+st+"\t"+ed+"\t"+strand+"\t"+type)
            c = "transcript_id"
            g = "gene_name"
            t = "transcript_type"
            getquote(line,outF,c)
            getquote(line,outF,g)
            getquote(line,outF,t)
            outF.write("\n")
outF.close() 
```
***
<a name="Home">[GoTop](#)</a>
<a name="Loading Data"></a>\ 

#### __Loading Data__

  In order to load your data from a local drive, use the following format.
  __Note__ that the `"PATH_FILE"` refers to the location of the summary data
  from the above section. For more details on how to load datasets click
  [here](https://support.rstudio.com/hc/en-us/articles/218611977-Importing-Data-with-RStudio)

##### loading data from local drive
 > data <- read.delim("PATH_FILE", comment.char = "#")

  Internal NPS package data is included in the data folder. The data can be
  loaded into R working console using `data()` function. Here we show an example
  of how to load a dataset `gencode` from the data directory. A quick view of
  the data can be achieved using `head(gencode)`.

```{r}
#library("GenomeInfoDb")
library("GenomicRanges")
#library("IRanges")
#library("NPS")
data("gencode")
#head(gencode)
```
<a name="Home">[GoTop](#)</a>

<a name="Filtering chr21 gencode"></a>\ 

#### __Filtering chr21__

  Here we show an extraction of "gencode" dataset using R commands. Note
  to replace `PATH_FILE` with file direcotry path. `gtf` refers to the full 
  genome file. A subset function was used to filter chr21 dataset as follows.

`chr21 <- subset(gencode, seqnames == "chr21")` #"genecode" = whole genome gtf

    > gtf = read.table("PATH_FILE")
    > gtf = subset(gtf, biotype == "transcript")
    > colnames(gtf) = c("chr","start","end","strand","biotype")
    > gr = GRanges(gtf)

<a name="Home">[GoTop](#)</a> 
<a name="Examples"></a>\

#### __Examples__

##### Processing Query
***

```{r}
query <- GRanges(c("chr1:2-10:+","chr1:6-10:-"), Row.names = c("trans1","trans2"),score = c(1,2))
head(query)

```

##### Classifying Biotypes
***

```{r}
#library(NPS)
#library("GenomeInfoDb")
#library("GenomicRanges")
#library("IRanges")
gr <- GRanges(c("chr1:1-5:+","chr1:2-3:+"),biotype = c("lincRNA","CPC"))
head(gr)
```

##### Extracting intron coordinates
*** 

      # Intron coordinates
      
       intron <- GRanges("chr1:6-8:+")
  
```{r}
#library(NPS)
#library("GenomeInfoDb")
#library("GenomicRanges")
#library("IRanges")

intron <- GRanges("chr1:6-8:+")
head(intron)
```

##### Filtering coding transcripts
***

    # Filtering non-coding regions using products from example 1, 2 and 3

```{r}
#library("NPS")
#library("GenomeInfoDb")
#library("GenomicRanges")
#library("IRanges")

#intron_trncp <- getBiotypes(query, gr, intron)
#intron_trncp
```

    # Filtering Intron and Exons 

  Here we show how to obtain protien coding and non-coding from our datasets.
  The coding transcrptis are an expressed sections of the genome that is
  responsibe for protein fomation. Meanwhile the non-coding transcripts are
  vital in the fromation reguraltory elements such promoters, enhancers and
  silencers.

```{r}
#library("NPS")
library("GenomeInfoDb")
library("GenomicRanges")
library("IRanges")
#data("intron")
#data("ILEF")
#data("gencode")
#gencode_gr = GRanges(gencode)
#ILEF_gr = GRanges(ILEF)
#cod_gr = GRanges(cod)
#intron_gr = GRanges(intron)

#non-coding <- getBiotypes(ILEF_gr, gencode_gr, intron_gr)
#non-coding
#
#coding <- getBiotypes(ILEF_gr, gencode_gr)
#coding
```


##### Finding overlapping transcripts
***
    # Samples with overlapping coding regions.
```{r}
#library("NPS")
#library("GenomicRanges")
#library("IRanges")
#library("GenomeInfoDb")

data("ILEF")
data("cod")
#ILEF_gr = GRanges(ILEF)
#cod_gr = GRanges(cod)

#rdthrough <- getReadthrough(ILEF_gr, cod_gr)
#rdthrough
```

#### __An Identification of Critical Tipping Point__
#### __Data Preprocessing__
#### __Pre-selection Transcript__
#### __Network Partition__
#### __Identifying Dynamic Network Biomodule__
#### __Finding Tipping Point__

  Next we will filter clinical data and group the transcripts into five
  different stages of disease progression: `"activated"`,
  `"lymphoma_aggressive"`, `"lymphoma_marginal"`, `"lymphoma_transitional,"` and
  `"resting"` as disease stage names. These stages resemble the general disease
  progression patterns: normal(healthy) state, predisease state, tipping point,
  critical transition stage(CTS) and disease stage. Currently it is vey
  difficult to identify when a disease progress from predisease state to
  critical transition state. In most cases these transitions occurs very fast.
  Ultimately we want to identify this critical tipping point before it
  progresses into a early disease stage.

  An existing dataset, GSE6136, is used to demonestrate how our functions are
  applied. Start by loading packages "stringr", "psych", "igraph" and "cluster".
  If these packages are not installed, use the `install.packages("packages")`
  function to install them. Then load them using `library("package")`. Below are
  some examples.

```{r}
#install.packages("stringr")
library("stringr")
library("psych")
library("igraph")
library("cluster")
```

  Once all the required packages are installed, load using "read.table()"
  function as follows. Note to change the `read.table(PATH/TO/YOUR/FILE)` when
  running your datasets. To check the dimension of your data set use
  "dim(dataset)" function. Here we show a use of dim function "dim(GSE6136)".
  Notice that after editing the column and row, the dimension of our dataset
  changes from (22690,27) to (22690, 26) because we removed the downloaded first
  row after assigning it to be column-name of the final numeric data matrix.

```{r}
source("C:/Users/benjk/Desktop/NPS/R/NPS.R")
GSE6136 = read.table("C:/Users/benjk/Desktop/GSE6136_matrix.txt", header = TRUE, comment = "!")

dim(GSE6136)               #[1] 22690rows and 27 columns
row.names(GSE6136) = GSE6136$ID_REF
GSE6136 = GSE6136[,-1]
dim(GSE6136)               #[1] 22690 rows and 26 columns
```

  The summary of GSE6136 dataset is shown below.

```{r}
cli = read.delim("C:/Users/benjk/Desktop/GSE6136_cli.txt", head = F)
dim(cli)
cli = t(cli)
colnames(cli) = str_split_fixed(cli[1,],'_',2)[,2]
cli = cli[-1,]
cli = data.frame(cli)
cli[,"cell-type:ch1"] = str_split_fixed(cli$characteristics_ch1.1,": ",2)[,2]
cli[,"Ig clonality:ch1"] = str_split_fixed(cli$characteristics_ch1.3,": ",2)[,2]

colnames(cli)[colnames(cli) == "cell-type:ch1"] = "group"
cli$Row.names = cli[,1]
head(cli)
```

  We normalized the expression of genes using log2() scale. This normalization 
  will ensure a more accurate comparison of the variance between the expression
  groups (clusters).

```{r}
dat <- GSE6136
df <- log2(dat+1)
head(df)
```

  After normailization, we now show how to classify diffrernt stages. The
  tipping point which is within the "activated" state in this case. Here we see
  the number of samples that are clasified into states "activated",
  "lymphoma_aggressive", "lymphoma_marginal", "lymphoma_transitional" and
  "resting". For instance, states "activated" and "resting" contain three and
  four samples, respectively. All the contents of the data set "test" can be 
  viewed using `View(test)`. Each clinical state's content can be viewed using
  `head(test["stage_name"])`. For instance, head(test["activated"]) shows
  contents of the activated state.

```{r}
tmp <- names(table(cli$group))
samplesL <- split(cli[,1],f = cli$group)
head(samplesL)
test <- sd_selection(df, samplesL,0.01)
head(test[["activated"]])
```

  A graphical represetation of genes of interest can be achieved using the
  functions shown below. The `getNetwork` function will obtain an igraph object
  based on a pearson correlation of `test`. This `igraphL` object is then run
  using the `getCluster_methods` function classify nodes.

```{r}
igraphL <- getNetwork(test, fdr = 0.05)
#summary(igraphL)
clstr <- getCluster_methods(igraphL)
#summary(clstr)
head(clstr)
```

  Finally, we can view a graph of classified clustered samples for five
  different stages. Use `"head(maxCIms)"` to view the CI scores calculated.

```{r}
#par(mfrow <- c(1,5))
#membersL_noweight <- getMCI(cluster, test, adjust.size = F, ylim = c(0,8))
#head(getMCI)
#maxCIms <- getMaxCImember(membersL_noweight[[1]],membersL_noweight[[2]],min =3)
#head(maxCIms)
```

<a name="">[GoTop](#)</a>
<a name="Acknowledgements"></a>\

#### __Acknowledgements__

  The development of this package would not be possible without continous help
  and feedback from individuals and institutions including: The Bioconductor
  Core Team, Dr. Xianan H Yang, Dr. Tzintzuni Garcia, and National Institutes of
  Health R21LM012619.

<a name="">[GoTop](#)</a>
<a name="SessionInfo"></a>\

```{r SessionInfo}
sessionInfo()
```
<a name="">[GoTop](#)</a>
<a name="References"></a>\

#### __References__ 
1. Scheffer M, Carpenter SR, Lenton TM, Bascompte J, Brock W, Dakos V, et al. Anticipating critical transitions. Science. 2012;338(6105):344-8. doi: 10.1126/science.1225244. PubMed PMID: 23087241.
2. Chen L, Liu R, Liu ZP, Li M, Aihara K. Detecting early-warning signals for sudden deterioration of complex diseases by dynamical network biomarkers. Sci Rep. 2012;2:342. Epub 2012/03/31. doi: 10.1038/srep00342. PubMed PMID: 22461973; PubMed Central PMCID: PMC3314989.
3. Mojtahedi M, Skupin A, Zhou J, Castano IG, Leong-Quong RY, Chang H, et al. Cell Fate Decision as High-Dimensional Critical State Transition. PLoS Biol. 2016;14(12):e2000640. doi: 10.1371/journal.pbio.2000640. PubMed PMID: 28027308; PubMed Central PMCID: PMCPMC5189937.
4. Wang ZZ, Cunningham JM, Yang XH. CisPi: a transcriptomic score for disclosing cis-acting disease-associated lincRNAs. Bioinformatics. 2018;34(17):664-70. doi: 10.1093/bioinformatics/bty574. PubMed PMID: WOS:000444317200009.



